### PR TITLE
Modifications to LZ4 iterable

### PR DESCRIPTION
**Pull request by Bouncner**
When not erasing LZ4 iterables, it turned out that certain std algorithms cannot use the iterators. With @mrks 's help, these issues have been fixed.  This change further allows to use unerased LZ4 segments, which are of relevance for my compression selection work and @krichly's spatial work.

Further, it adds a slight change to the caching of blocks. With this change, an iterable that is accessed multiple times, can still profit from LZ4-block caching (will be relevant for upcoming changes to the segment accessors).

TPC-H (gcc, sf=1, nemea):
```diff
 +----------------+-----------------------+------+----------------------+------+------------+-------------------+
 | Benchmark      | prev. iter/s          | runs | new iter/s           | runs | change [%] |           p-value |
 +----------------+-----------------------+------+----------------------+------+------------+-------------------+
 | TPC-H 01       | 0.0077946605160832405 | 1    | 0.007878044620156288 | 1    | +1%        | (not enough runs) |
 | TPC-H 02       | 0.15306907892227173   | 10   | 0.1545109897851944   | 10   | +1%        |            0.0003 |
 | TPC-H 03       | 0.1601416915655136    | 10   | 0.16082912683486938  | 10   | +0%        |            0.0161 |
 | TPC-H 04       | 0.3642721176147461    | 22   | 0.37286096811294556  | 23   | +2%        |            0.0000 |
 | TPC-H 05       | 0.03404029458761215   | 3    | 0.033995673060417175 | 3    | -0%        | (not enough runs) |
+| TPC-H 06       | 1.9570527076721191    | 118  | 2.0897433757781982   | 126  | +7%        |            0.0069 |
 | TPC-H 07       | 0.16901127994060516   | 11   | 0.16961240768432617  | 11   | +0%        |            0.0014 |
 | TPC-H 08       | 0.15593402087688446   | 10   | 0.1585163027048111   | 10   | +2%        |            0.0000 |
 | TPC-H 09       | 0.026666978374123573  | 2    | 0.0267617329955101   | 2    | +0%        | (not enough runs) |
 | TPC-H 10       | 0.1189955472946167    | 8    | 0.12087590247392654  | 8    | +2%        | (not enough runs) |
 | TPC-H 11       | 0.4064975678920746    | 25   | 0.41653311252593994  | 25   | +2%        |            0.0000 |
 | TPC-H 12       | 0.05560759827494621   | 4    | 0.05629827082157135  | 4    | +1%        | (not enough runs) |
 | TPC-H 13       | 0.223331481218338     | 14   | 0.22616297006607056  | 14   | +1%        |            0.0000 |
 | TPC-H 14       | 1.095063328742981     | 66   | 1.1197049617767334   | 68   | +2%        |            0.0000 |
 | TPC-H 15       | 0.4771794378757477    | 29   | 0.47672492265701294  | 29   | -0%        |            0.7583 |
 | TPC-H 16       | 0.08077448606491089   | 5    | 0.0832173302769661   | 5    | +3%        | (not enough runs) |
 | TPC-H 17       | 0.021238891407847404  | 2    | 0.021426254883408546 | 2    | +1%        | (not enough runs) |
 | TPC-H 18       | 0.018045440316200256  | 2    | 0.018120162189006805 | 2    | +0%        | (not enough runs) |
+| TPC-H 19       | 0.3555009961128235    | 22   | 0.462495356798172    | 28   | +30%       |            0.0000 |
 | TPC-H 20       | 0.059324923902750015  | 4    | 0.05963096022605896  | 4    | +1%        | (not enough runs) |
 | TPC-H 21       | 0.012740414589643478  | 1    | 0.012884444557130337 | 1    | +1%        | (not enough runs) |
 | TPC-H 22       | 0.15871000289916992   | 10   | 0.15976382791996002  | 10   | +1%        |            0.0000 |
 | geometric mean |                       |      |                      |      | +3%        |                   |
 +----------------+-----------------------+------+----------------------+------+------------+-------------------+
```

### CODE REVIEW

File: src/lib/storage/lz4_segment/lz4_segment_iterable.hpp [hunk 1]
```diff
@@ -19,18 +19,15 @@ class LZ4SegmentIterable : public PointAccessibleSegmentIterable<LZ4SegmentItera
   template <typename Functor>
   void _on_with_iterators(const Functor& functor) const {
     using ValueIterator = typename std::vector<T>::const_iterator;
+    using NullValueIterator = typename pmr_vector<bool>::const_iterator;
```
[the remark this answers is not in the document]
Comment (Bouncner) [reply src/lib/storage/lz4_segment/lz4_segment_iterable.hpp]: For the current TPC-H, it's not much. As soon as a query on a sorted segment is executed, it's a lot (can be factors, depending on input size, selectivity etc.). For a reason I do not understand, `AnySegmentIterables` do not support binary searches (they should be random accessible though).

With the change to the segment accessors (I actually don't dare to say it will end up in the master, as they main advantage is for caching iterators and that's only LZ4), it makes again more sense as iterables are reused much more often.

So in short: currently not much, later maybe more. :)
So always erasing them is still the correct decision imo.

File: src/lib/storage/lz4_segment/lz4_segment_iterable.hpp [hunk 2]
```diff
@@ -19,18 +19,15 @@ class LZ4SegmentIterable : public PointAccessibleSegmentIterable<LZ4SegmentItera
...
-     * we can just use a default initialized boolean vector.
-     */
-    const auto null_values = _segment.null_values() ? *_segment.null_values() : pmr_vector<bool>(_segment.size());
-
-    auto begin = Iterator<ValueIterator>{decompressed_segment.cbegin(), null_values.cbegin()};
-    auto end = Iterator<ValueIterator>{decompressed_segment.cend(), null_values.cend()};
-
+    std::optional<NullValueIterator> null_opt_value_iter;
+    auto begin = Iterator<ValueIterator>{
+        decompressed_segment.cbegin(), _segment.null_values() ? _segment.null_values()->cbegin() : null_opt_value_iter};
```
[the remark this answers is not in the document]
Comment (Bouncner) [reply src/lib/storage/lz4_segment/lz4_segment_iterable.hpp]: Will the `null_values` iterator be copied here in case it exists?

[the remark this answers is not in the document]
Comment (mrks) [reply src/lib/storage/lz4_segment/lz4_segment_iterable.hpp]: You create it here as part of the method call, so it's a prvalue. Currently, you copy it from the `const&` to the local member as part of your iterator constructor. You could avoid that by accepting it by value and moving it into the member.

[the remark this answers is not in the document]
Comment (Bouncner) [reply src/lib/storage/lz4_segment/lz4_segment_iterable.hpp]: Is there a way around that error?
```../src/lib/storage/lz4_segment/lz4_segment_iterable.hpp:26:63: error: incompatible operand types ('std::__1::vector<bool, boost::container::pmr::polymorphic_allocator<bool> >::const_iterator' (aka '__bit_iterator<std::__1::vector<bool, boost::container::pmr::polymorphic_allocator<bool> >, true>') and 'const std::__1::nullopt_t')
        decompressed_segment.cbegin(), _segment.null_values() ? _segment.null_values()->cbegin() : std::nullopt};```

that is not an additional wrap like this `std::optional<NullValueIterator>{_segment.null_values()}`????

And is the move really necessary (adding any advantage) here?

[the remark this answers is not in the document]
Comment (mrks) [reply src/lib/storage/lz4_segment/lz4_segment_iterable.hpp]: Instead of `std::nullopt`, try to use `std::optional<NullValueIterator>{}`. Sorry.

> And is the move really necessary (adding any advantage) here?

No idea, but when a constructor takes a complex object by value, moving it is more or less the standard.

[the remark this answers is not in the document]
Comment (Bouncner) [reply src/lib/storage/lz4_segment/lz4_segment_iterable.hpp]: No worries. As often as I am positively surprised what compilers are capable of, as often I am totally confused why simple things cannot be resolved. Will add the move.

File: src/lib/storage/lz4_segment/lz4_segment_iterable.hpp [hunk 3]
```diff
@@ -42,22 +39,28 @@ class LZ4SegmentIterable : public PointAccessibleSegmentIterable<LZ4SegmentItera
...
+    const auto position_filter_size = position_filter->size();
+
+    auto decompressed_filtered_segment = std::vector<ValueType>(position_filter_size);
+    for (auto index = size_t{0u}; index < position_filter_size; ++index) {
       const auto& position = (*position_filter)[index];
       // NOLINTNEXTLINE
       auto [value, block_index] = _segment.decompress(position.chunk_offset, cached_block_index, cached_block);
       decompressed_filtered_segment[index] = std::move(value);
       cached_block_index = block_index;
     }
```
[the remark this answers is not in the document]
Comment (Bouncner) [reply src/lib/storage/lz4_segment/lz4_segment_iterable.hpp]: Added a comment.

**Comment by Bouncner** [comment]
Feedback so far should be addressed. Benchmark is running.

File: src/lib/storage/lz4_segment/lz4_segment_iterable.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (Bouncner) [reply src/lib/storage/lz4_segment/lz4_segment_iterable.hpp]: Changed. Declaring the type with the nullopt was shorter than what I had before.

File: src/lib/storage/lz4_segment/lz4_segment_iterable.hpp [hunk 4]
```diff
@@ -42,21 +40,34 @@ class LZ4SegmentIterable : public PointAccessibleSegmentIterable<LZ4SegmentItera
...
-    auto decompressed_filtered_segment = std::vector<ValueType>(position_filter->size());
-    auto cached_block = std::vector<char>{};
-    auto cached_block_index = std::optional<size_t>{};
-    for (auto index = size_t{0u}; index < position_filter->size(); ++index) {
+    // vector storing the uncompressed values
+    auto decompressed_filtered_segment = std::vector<ValueType>(position_filter_size);
+
+    // _segment.decompress() takes the currently cached block (reference) and its id in addition to the requested
+    // element. If the requested element in not within that block, the next block will be decompressed and written to
+    // `cached_block` while the value and the new block id are returned. In case the requested element in within the
```
Comment (mrks) [inline src/lib/storage/lz4_segment/lz4_segment_iterable.hpp]: typo: is

File: src/lib/storage/lz4_segment/lz4_segment_iterable.hpp [hunk 5]
```diff
@@ -42,21 +40,34 @@ class LZ4SegmentIterable : public PointAccessibleSegmentIterable<LZ4SegmentItera
...
-                                                    position_filter->cbegin(), position_filter->cbegin()};
-    auto end = PointAccessIterator<ValueIterator>{decompressed_filtered_segment, &_segment.null_values(),
+    auto begin = PointAccessIterator<ValueIterator>{
+        decompressed_filtered_segment.begin(),
+        _segment.null_values() ? std::optional<NullValueIterator>{_segment.null_values()->cbegin()}
+                               : std::optional<NullValueIterator>{},
+        position_filter->cbegin(), position_filter->cbegin()};
+    auto end = PointAccessIterator<ValueIterator>{decompressed_filtered_segment.begin(),
+                                                  _segment.null_values()
+                                                      ? std::optional<NullValueIterator>{_segment.null_values()->cend()}
```
Comment (mrks) [inline src/lib/storage/lz4_segment/lz4_segment_iterable.hpp]: I'm surprised that you need the std::optional here

File: src/lib/storage/lz4_segment/lz4_segment_iterable.hpp [hunk 6]
```diff
@@ -77,35 +90,28 @@ class LZ4SegmentIterable : public PointAccessibleSegmentIterable<LZ4SegmentItera
 
    public:
     // Begin and End Iterator
-    explicit Iterator(ValueIterator data_it, const NullValueIterator null_value_it)
-        : _chunk_offset{0u}, _data_it{data_it}, _null_value_it{null_value_it} {}
+    explicit Iterator(ValueIterator data_it, const std::optional<NullValueIterator> null_value_it)
+        : _chunk_offset{0u}, _data_it{std::move(data_it)}, _null_value_it{std::move(null_value_it)} {}
```
Comment (mrks) [inline src/lib/storage/lz4_segment/lz4_segment_iterable.hpp]: Surprised that clang-tidy doesn't complain about this: You cannot move from a const.

File: src/lib/storage/lz4_segment/lz4_segment_iterable.hpp [hunk 7]
```diff
@@ -128,29 +136,31 @@ class LZ4SegmentIterable : public PointAccessibleSegmentIterable<LZ4SegmentItera
    public:
     using ValueType = T;
     using IterableType = LZ4SegmentIterable<T>;
+    using DataIteratorType = typename std::vector<T>::const_iterator;
+    using NullValueIterator = typename pmr_vector<bool>::const_iterator;
 
     // Begin Iterator
-    PointAccessIterator(const std::vector<T>& data, const std::optional<pmr_vector<bool>>* null_values,
-                        const PosList::const_iterator position_filter_begin, PosList::const_iterator position_filter_it)
+    PointAccessIterator(DataIteratorType data_it, const std::optional<NullValueIterator>& null_value_it,
```
Comment (mrks) [inline src/lib/storage/lz4_segment/lz4_segment_iterable.hpp]: same: just `std::optional<NullValueIterator>`

File: src/lib/storage/lz4_segment/lz4_segment_iterable.hpp (hunk 6 above)
Comment (Bouncner) [reply src/lib/storage/lz4_segment/lz4_segment_iterable.hpp]: remove the const?

File: src/lib/storage/lz4_segment/lz4_segment_iterable.hpp (hunk 7 above)
Comment (Bouncner) [reply src/lib/storage/lz4_segment/lz4_segment_iterable.hpp]: What? I removed that 5 times already.

File: src/lib/storage/lz4_segment/lz4_segment_iterable.hpp (hunk 5 above)
Comment (Bouncner) [reply src/lib/storage/lz4_segment/lz4_segment_iterable.hpp]: Narf ... I must have screwed something up. That was somewhere else.

Anyhow, does not compile if neither side of the `:` does have the `std::optional< NullValueIterator>`.

File: src/lib/storage/lz4_segment/lz4_segment_iterable.hpp (hunk 6 above)
Comment (mrks) [reply src/lib/storage/lz4_segment/lz4_segment_iterable.hpp]: yes

File: src/lib/storage/lz4_segment/lz4_segment_iterable.hpp (hunk 5 above)
Comment (mrks) [reply src/lib/storage/lz4_segment/lz4_segment_iterable.hpp]: ok

**Comment by Bouncner** [comment]
I think @mrks's second round of issues should be addressed.

File: src/lib/storage/lz4_segment/lz4_segment_iterable.hpp (hunk 4 above)
Comment (mrks) [reply src/lib/storage/lz4_segment/lz4_segment_iterable.hpp]: Any chance you forgot to push something? This is still unadressed.

File: src/lib/storage/lz4_segment/lz4_segment_iterable.hpp [hunk 8]
```diff
@@ -42,21 +40,34 @@ class LZ4SegmentIterable : public PointAccessibleSegmentIterable<LZ4SegmentItera
...
-    auto end = PointAccessIterator<ValueIterator>{decompressed_filtered_segment, &_segment.null_values(),
+    auto begin = PointAccessIterator<ValueIterator>{
+        decompressed_filtered_segment.begin(),
+        _segment.null_values() ? _segment.null_values()->cbegin()
+                               : std::optional<NullValueIterator>{},
+        position_filter->cbegin(), position_filter->cbegin()};
+    auto end = PointAccessIterator<ValueIterator>{decompressed_filtered_segment.begin(),
+                                                  _segment.null_values()
+                                                      ? _segment.null_values()->cend()
+                                                      : std::optional<NullValueIterator>{},
```
Comment (mrks) [inline src/lib/storage/lz4_segment/lz4_segment_iterable.hpp]: This is... not really nice. How about we just wrap it into an `if (_segment.null_values())`?

File: src/lib/storage/lz4_segment/lz4_segment_iterable.hpp (hunk 4 above)
Comment (Bouncner) [reply src/lib/storage/lz4_segment/lz4_segment_iterable.hpp]: You mean the "is" needed in "If the requested element is not within that block", don't you?

File: src/lib/storage/lz4_segment/lz4_segment_iterable.hpp (hunk 8 above)
Comment (Bouncner) [reply src/lib/storage/lz4_segment/lz4_segment_iterable.hpp]: You mean empty begin/end and than set them depending on `_segment.null_values()`? Not much better IMO. Or do you mean something else?

File: src/lib/storage/lz4_segment/lz4_segment_iterable.hpp (hunk 4 above)
Comment (mrks) [reply src/lib/storage/lz4_segment/lz4_segment_iterable.hpp]: In case the requested element **in**

File: src/lib/storage/lz4_segment/lz4_segment_iterable.hpp (hunk 8 above)
Comment (Bouncner) [reply src/lib/storage/lz4_segment/lz4_segment_iterable.hpp]: Jep ... MUCH better.

**Comment by mrks** [comment]
> @tobodner self-requested a review 23 hours ago

Do you still want a chance to look at this?

**Comment by tobodner** [comment]
> > @tobodner self-requested a review 23 hours ago
> 
> Do you still want a chance to look at this?

I do not want to block submission. I am happy to look at this and learn from it post commit.
